File: examples/python/cart/handlers/checkout.py
"""Checkout command handler."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains
from .state import CartState

from .errors import CommandRejectedError


def handle_checkout(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Cart does not exist")
    if not state.is_active():
        raise CommandRejectedError("Cart is already checked out")
    if not state.items:
        raise CommandRejectedError("Cart is empty")

    log.info("checking_out")

    event = domains.CartCheckedOut(
        final_subtotal=state.subtotal_cents,
        discount_cents=state.discount_cents,
        checked_out_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/rust/cart/src/handlers/checkout.rs
//! Checkout command handler.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CartCheckedOut, CartState};
use common::{now, require_exists, require_not_empty, require_status_not, Result};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the Checkout command.
///
/// Finalizes the cart and marks it as checked out.
pub fn handle_checkout(
    command_book: &CommandBook,
    _command_data: &[u8],
    state: &CartState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.customer_id, errmsg::CART_NOT_FOUND)?;
    require_status_not(&state.status, "checked_out", errmsg::CART_CHECKED_OUT)?;
    require_not_empty(&state.items, errmsg::CART_EMPTY)?;

    let event = CartCheckedOut {
        final_subtotal: state.subtotal_cents,
        discount_cents: state.discount_cents,
        checked_out_at: Some(now()),
        customer_id: state.customer_id.clone(),
        items: state.items.clone(),
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.CartCheckedOut",
        event,
    ))
}
